File: implementations/cpp/dynamic_programming_intuition.cpp
#include <cassert>
#include <cstdint>
#include <iostream>
#include <vector>
#include <algorithm>

namespace dsa {

/**
 * @brief Archetype 1: Fibonacci Sequence.
 * Demonstrates Naive Recursion vs Memoization vs Tabulation vs Space Optimization.
 */
class FibonacciDemo {
public:
    // 1. Naive Recursion: O(2^N) time, O(N) stack space
    static uint64_t naive_recursive(int n) {
        if (n <= 1) return static_cast<uint64_t>(n);
        return naive_recursive(n - 1) + naive_recursive(n - 2);
    }

    // 2. Top-Down Memoization: O(N) time, O(N) space
    static uint64_t memoized(int n, std::vector<uint64_t>& memo) {
        if (n <= 1) return static_cast<uint64_t>(n);
        if (memo[n] != 0) return memo[n];
        return memo[n] = memoized(n - 1, memo) + memoized(n - 2, memo);
    }

    // 3. Bottom-Up Tabulation: O(N) time, O(N) table space
    static uint64_t tabulated(int n) {
        if (n <= 1) return static_cast<uint64_t>(n);
        std::vector<uint64_t> dp(n + 1, 0);
        dp[1] = 1;
        for (int i = 2; i <= n; ++i) {
            dp[i] = dp[i - 1] + dp[i - 2];
        }
        return dp[n];
    }

    // 4. Space-Optimized Tabulation: O(N) time, O(1) space
    static uint64_t space_optimized(int n) {
        if (n <= 1) return static_cast<uint64_t>(n);
        uint64_t prev2 = 0, prev1 = 1;
        for (int i = 2; i <= n; ++i) {
            uint64_t curr = prev1 + prev2;
            prev2 = prev1;
            prev1 = curr;
        }
        return prev1;
    }
};

/**
 * @brief Archetype 2: 2D Grid Unique Paths.
 * Evaluates state transition on a Directed Acyclic Graph (DAG).
 */
class GridPathsDemo {
public:
    // Tabulation with 2D array: O(M * N) time, O(M * N) space
    static int64_t unique_paths_2d(int m, int n) {
        std::vector<std::vector<int64_t>> dp(m, std::vector<int64_t>(n, 1));
        for (int i = 1; i < m; ++i) {
            for (int j = 1; j < n; ++j) {
                dp[i][j] = dp[i - 1][j] + dp[i][j - 1];
            }
        }
        return dp[m - 1][n - 1];
    }

    // Space-optimized with 1D rolling array: O(M * N) time, O(N) space
    static int64_t unique_paths_1d(int m, int n) {
        std::vector<int64_t> dp(n, 1);
        for (int i = 1; i < m; ++i) {
            for (int j = 1; j < n; ++j) {
                dp[j] += dp[j - 1];
            }
        }
        return dp[n - 1];
    }
};

/**
 * @brief Archetype 3: 0-1 Knapsack Problem.
 * Demonstrates Optimal Substructure and 1D reverse-iteration rolling array.
 */
class KnapsackDemo {
public:
    static int knapsack_1d(const std::vector<int>& weights,
                           const std::vector<int>& values,
                           int capacity) {
        std::vector<int> dp(capacity + 1, 0);
        for (size_t i = 0; i < weights.size(); ++i) {
            int w = weights[i];
            int v = values[i];
            // Reverse iteration ensures each item is used at most once
            for (int cap = capacity; cap >= w; --cap) {
                dp[cap] = std::max(dp[cap], dp[cap - w] + v);
            }
        }
        return dp[capacity];
    }
};

} // namespace dsa

int main() {
    using namespace dsa;

    std::cout << "[RUNNING] Dynamic Programming Intuition C++17 Verification..." << std::endl;

    // 1. Fibonacci Equivalence
    {
        int n = 15;
        uint64_t ans_naive = FibonacciDemo::naive_recursive(n);
        std::vector<uint64_t> memo(n + 1, 0);
        uint64_t ans_memo = FibonacciDemo::memoized(n, memo);
        uint64_t ans_tab = FibonacciDemo::tabulated(n);
        uint64_t ans_opt = FibonacciDemo::space_optimized(n);

        assert(ans_naive == 610);
        assert(ans_memo == ans_naive);
        assert(ans_tab == ans_naive);
        assert(ans_opt == ans_naive);

        // Larger n where naive would timeout
        assert(FibonacciDemo::space_optimized(50) == 12586269025ULL);
    }

    // 2. Grid Paths
    {
        assert(GridPathsDemo::unique_paths_2d(3, 7) == 28);
        assert(GridPathsDemo::unique_paths_1d(3, 7) == 28);
        assert(GridPathsDemo::unique_paths_1d(1, 1) == 1);
        assert(GridPathsDemo::unique_paths_1d(3, 3) == 6);
    }

    // 3. 0-1 Knapsack
    {
        std::vector<int> weights = {1, 3, 4, 5};
        std::vector<int> values = {1, 4, 5, 7};
        int capacity = 7;
        // Best: item 1 (w=3, v=4) + item 2 (w=4, v=5) -> w=7, v=9
        assert(KnapsackDemo::knapsack_1d(weights, values, capacity) == 9);
    }

    std::cout << "[PASSED] Dynamic Programming Intuition C++17 All Tests Passed!" << std::endl;
    return 0;
}

File: implementations/python/dynamic_programming_intuition.py
"""
Dynamic Programming Intuition implementations in Python.

Demonstrates:
1. Fibonacci: Naive Recursion vs Top-Down Memoization vs Bottom-Up Tabulation vs Space Optimization.
2. 2D Grid Unique Paths: Full 2D Table vs 1D Rolling Array.
3. 0-1 Knapsack: Optimal Substructure & 1D reverse-iteration table.
4. Comprehensive unit tests.
"""

import unittest
from typing import List, Dict


def fib_naive(n: int) -> int:
    """Naive recursion: O(2^n) time, O(n) stack space."""
    if n <= 1:
        return n
    return fib_naive(n - 1) + fib_naive(n - 2)


def fib_memo(n: int, memo: Dict[int, int] = None) -> int:
    """Top-down memoization: O(n) time, O(n) space."""
    if memo is None:
        memo = {}
    if n <= 1:
        return n
    if n not in memo:
        memo[n] = fib_memo(n - 1, memo) + fib_memo(n - 2, memo)
    return memo[n]


def fib_tab(n: int) -> int:
    """Bottom-up tabulation: O(n) time, O(n) table space."""
    if n <= 1:
        return n
    dp = [0] * (n + 1)
    dp[1] = 1
    for i in range(2, n + 1):
        dp[i] = dp[i - 1] + dp[i - 2]
    return dp[n]


def fib_opt(n: int) -> int:
    """Space-optimized tabulation: O(n) time, O(1) space."""
    if n <= 1:
        return n
    prev2, prev1 = 0, 1
    for _ in range(2, n + 1):
        curr = prev1 + prev2
        prev2 = prev1
        prev1 = curr
    return prev1


def grid_paths_2d(m: int, n: int) -> int:
    """Unique paths in m x n grid using full 2D table: O(m * n) time and space."""
    dp = [[1] * n for _ in range(m)]
    for i in range(1, m):
        for j in range(1, n):
            dp[i][j] = dp[i - 1][j] + dp[i][j - 1]
    return dp[m - 1][n - 1]


def grid_paths_1d(m: int, n: int) -> int:
    """Unique paths in m x n grid using 1D rolling array: O(m * n) time, O(n) space."""
    dp = [1] * n
    for _ in range(1, m):
        for j in range(1, n):
            dp[j] += dp[j - 1]
    return dp[n - 1]


def knapsack_1d(weights: List[int], values: List[int], capacity: int) -> int:
    """0-1 Knapsack using 1D reverse-iteration table: O(n * W) time, O(W) space."""
    dp = [0] * (capacity + 1)
    for w, v in zip(weights, values):
        for cap in range(capacity, w - 1, -1):
            dp[cap] = max(dp[cap], dp[cap - w] + v)
    return dp[capacity]


class TestDynamicProgrammingIntuition(unittest.TestCase):
    def test_fibonacci(self):
        n = 15
        val_naive = fib_naive(n)
        val_memo = fib_memo(n)
        val_tab = fib_tab(n)
        val_opt = fib_opt(n)

        self.assertEqual(val_naive, 610)
        self.assertEqual(val_memo, 610)
        self.assertEqual(val_tab, 610)
        self.assertEqual(val_opt, 610)
        self.assertEqual(fib_opt(50), 12586269025)

    def test_grid_paths(self):
        self.assertEqual(grid_paths_2d(3, 7), 28)
        self.assertEqual(grid_paths_1d(3, 7), 28)
        self.assertEqual(grid_paths_1d(1, 1), 1)
        self.assertEqual(grid_paths_1d(3, 3), 6)

    def test_knapsack(self):
        weights = [1, 3, 4, 5]
        values = [1, 4, 5, 7]
        capacity = 7
        self.assertEqual(knapsack_1d(weights, values, capacity), 9)


if __name__ == '__main__':
    unittest.main()
